Make source links look cleaner

jsha opened the pull request:
Change from syntaxy-looking [src] to the plain word "source".

Change the syntaxy-looking `[-]` at the top of the page to say "collapse".

Reduce opacity of rightside content.

Part of #59851

r? @GuillaumeGomez

Demo: https://rustdoc.crud.net/jsha/source-link-2/std/string/struct.String.html

[Discussed on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/display.20of.20source.20link).

rust-highfive commented:
Some changes occurred in HTML/CSS/JS.

cc @GuillaumeGomez,@Folyd

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4722229914?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/e06bbb09d064d64727b449d3b1f5dab02881e9e6/checks/4722229914/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Check compiletest suite=rustdoc mode=rustdoc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
warning: `tidy` is not installed; diffs will not be generated

running 490 tests
i....................................................i...........................................F.. 100/490
...F................................................................................................ 200/490
.....................FF....................................F........................................ 300/490
............................i.................................................................i..... 400/490
........................F..............................F.......F..........................
Some tests failed in compiletest suite=rustdoc mode=rustdoc host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu

---- [rustdoc] rustdoc/ensure-src-link.rs stdout ----


error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/ensure-src-link" "/checkout/src/test/rustdoc/ensure-src-link.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
5: @has check failed
 `XPATH PATTERN` did not match
 // @has foo/trait.Iterator.html '//div[@id="method.zip"]//a[@class="srclink"]' "[src]"
Encountered 1 errors

------------------------------------------



---- [rustdoc] rustdoc/external-macro-src.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/external-macro-src" "/checkout/src/test/rustdoc/external-macro-src.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
8: @has check failed
 `XPATH PATTERN` did not match
 // @has foo/index.html '//a[@href="../src/foo/external-macro-src.rs.html#3-12"]' '[src]'
11: @has check failed
 `XPATH PATTERN` did not match
 // @has - '//a[@href="../src/foo/external-macro-src.rs.html#12"]' '[src]'
Encountered 2 errors

------------------------------------------



---- [rustdoc] rustdoc/issue-16265-1.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/issue-16265-1" "/checkout/src/test/rustdoc/issue-16265-1.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
3: @has check failed
 `PATTERN` did not match
 // @has issue_16265_1/traits/index.html '[src]'
Encountered 1 errors

------------------------------------------



---- [rustdoc] rustdoc/issue-16265-2.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/issue-16265-2" "/checkout/src/test/rustdoc/issue-16265-2.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
1: @has check failed
 `PATTERN` did not match
 // @has issue_16265_2/index.html '[src]'
Encountered 1 errors

------------------------------------------



---- [rustdoc] rustdoc/issue-26606.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/issue-26606" "/checkout/src/test/rustdoc/issue-26606.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
10: @has check failed
 `XPATH PATTERN` did not match
 // @has - '//a[@href="../src/issue_26606/issue-26606.rs.html#11"]' '[src]'
Encountered 1 errors

------------------------------------------



---- [rustdoc] rustdoc/src-links-auto-impls.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/src-links-auto-impls" "/checkout/src/test/rustdoc/src-links-auto-impls.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
9: @has check failed
 `XPATH PATTERN` did not match
 // @has - '//div[@id="impl-Any"]//a[@class="srclink"]' '[src]'
Encountered 1 errors

------------------------------------------



---- [rustdoc] rustdoc/thread-local-src.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/thread-local-src" "/checkout/src/test/rustdoc/thread-local-src.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
3: @has check failed
 `XPATH PATTERN` did not match
 // @has foo/index.html '//a[@href="../src/foo/thread-local-src.rs.html#1-6"]' '[src]'
5: @has check failed
 `XPATH PATTERN` did not match
 // @has foo/constant.FOO.html '//a[@href="../src/foo/thread-local-src.rs.html#6"]' '[src]'
Encountered 2 errors

------------------------------------------



---- [rustdoc] rustdoc/trait-src-link.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/trait-src-link" "/checkout/src/test/rustdoc/trait-src-link.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
3: @has check failed
 `XPATH PATTERN` did not match
     // @has quix/trait.Foo.html '//a[@href="../src/quix/trait-src-link.rs.html#4"]' '[src]'
6: @has check failed
 `XPATH PATTERN` did not match
     // @has quix/trait.Foo.html '//a[@href="../src/quix/trait-src-link.rs.html#7"]' '[src]'
13: @has check failed
 `XPATH PATTERN` did not match
     // @has quix/struct.Bar.html '//a[@href="../src/quix/trait-src-link.rs.html#14"]' '[src]'
15: @has check failed
 `XPATH PATTERN` did not match
     // @has quix/struct.Bar.html '//a[@href="../src/quix/trait-src-link.rs.html#7"]' '[src]'
21: @has check failed
 `XPATH PATTERN` did not match
     // @has quix/struct.Baz.html '//a[@href="../src/quix/trait-src-link.rs.html#22"]' '[src]'
24: @has check failed
 `XPATH PATTERN` did not match
     // @has quix/struct.Baz.html '//a[@href="../src/quix/trait-src-link.rs.html#25"]' '[src]'
Encountered 6 errors

------------------------------------------

---
test result: FAILED. 478 passed; 8 failed; 4 ignored; 0 measured; 0 filtered out; finished in 46.11s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/rustdoc" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "rustdoc" "--mode" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:17:12
```

</details>

camelid commented:
Looks good! I don't see the "stable since" change though? Also, the color of "source" looks a little too pale in the light theme.

jsha commented:
> I don't see the "stable since" change though?

My original version added the "stable since" language but on reviewing the Zulip thread I saw there wasn't really consensus on that, and it's not the main point of the PR, so I reverted it (but forgot to update the PR description). I've now updated the PR description.

> Also, the color of "source" looks a little too pale in the light theme.

I'll bump it up a little.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4722678979?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/151a5331c05cb67ceac7a3c23a817e81d4f542c9/checks/4722678979/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  IMAGE: x86_64-gnu-tools
##[endgroup]
From https://github.com/rust-lang/rust
 * branch              master     -> FETCH_HEAD
Searching for toolstate changes between f1ce0e6a00593493a12e0e3662119786c761f375 and 5d7aa1612f9c88ec9b4a8cad00841a1ce0601984
Rustdoc was updated
##[group]Run src/ci/scripts/verify-channel.sh
src/ci/scripts/verify-channel.sh
shell: /bin/bash --noprofile --norc -e -o pipefail {0}
env:
---
.......... (50/54)
....       (54/54)


/checkout/src/test/rustdoc-gui/anchors.goml anchors... FAILED
[ERROR] (line 13) Error: Evaluation failed: expected `rgb(0, 0, 0)` for key `color` for selector `#toggle-all-docs`, found `rgb(56, 115, 173)`: for command `assert-css: ("#toggle-all-docs", {"color": "rgb(0, 0, 0)"})`



command did not execute successfully: "/node-v14.4.0-linux-x64/bin/node" "/checkout/src/tools/rustdoc-gui/tester.js" "--jobs" "16" "--doc-folder" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/doc" "--tests-folder" "/checkout/src/test/rustdoc-gui"


Build completed unsuccessfully in 0:00:16
```

</details>

camelid commented:
That looks better, although I think the links should be still deeper in color. The current color in some ways feels more distracting than if it were darker because it's on the edge of being less visible vs more visible. It just doesn't look right to me for whatever reason.

jsha commented:
Just pushed a version where they get normal opacity like any other link. Lemme know what you think.

camelid commented:
That looks better, thanks.

GuillaumeGomez commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -312,15 +312,15 @@ impl AllTypes {
+             <span class=\"out-of-band\">\
+                <span id=\"render-detail\">\
+                    <a id=\"toggle-all-docs\" href=\"javascript:void(0)\" \
+                    title=\"collapse all docs\">\
```
nit: could could indent the `title` attribute so my brain doesn't interpret it as a tag please?

GuillaumeGomez commented:
Apart from my comment, looks good to me.

GuillaumeGomez commented:
So looks good to me. I see some :-1: signs on the first comment.

@euclio @falaosu: What don't you like with this approach?

euclio commented:
My original objection was that the new text isn't visually distinct enough, though after reading the linked issue I see that there's a justification for this.

That said, I don't think that this does a good enough job of communicating that these are clickable actions. The brackets in the current style do a good job of making it clear that these are buttons, and "-" communicates the same amount of information as "collapse" in fewer characters. For such commonly-used actions that are repeated across the page, we can use symbols to communicate these actions effectively, and that reduces overall clutter.

"source" is maybe fine as a link visually, since it actually takes you to a separate page (though there's no underline on hover). I don't think collapse should just be a link, though, since it only performs an action on the current page. Assuming that we want to change the other "[-]"s on the page for consistency, I would expect a symbol like "▶︎" for each individual section, and "▼ collapse all" at the top.

jsha commented:
> "source" is maybe fine as a link visually, since it actually takes you to a separate page (though there's no underline on hover).

I'll add underline on hover.

> I don't think collapse should just be a link, though, since it only performs an action on the current page. Assuming that we want to change the other "[-]"s on the page for consistency, I would expect a symbol like "▶︎" for each individual section, and "▼ collapse all" at the top.

I changed `[-]` to `collapse` because `1.0.0 · [-] · source` looked wrong. But I could make it `1.0.0 · source · [-]`, in the spirit of "change one thing at a time."

GuillaumeGomez commented:
> "source" is maybe fine as a link visually, since it actually takes you to a separate page (though there's no underline on hover). I don't think collapse should just be a link, though, since it only performs an action on the current page. Assuming that we want to change the other "[-]"s on the page for consistency, I would expect a symbol like "▶︎" for each individual section, and "▼ collapse all" at the top.

To be fair, this button at the top of the page is a bit on its own since its behaviour is different than the other collapse toggles. But at least I see your point.

@jsha: Just realized that you didn't add a test for the text of the top collapse button (to check it's "expand"/"collapse" when it should). Could you add one as well please? And since you said you would add underline on hover, can you add a test for this as well please?

jsha commented:
Alright, I've pushed new code reverting the "collapse" text back to `[-]`. Updated the demo.

> Just realized that you didn't add a test for the text of the top collapse button (to check it's "expand"/"collapse" when it should).

Since this is reverted, no behavior here has changed. I'll look at adding a test for the `+` / `-` change though.

> And since you said you would add underline on hover, can you add a test for this as well please?

Will do.

camelid commented:
FWIW, I liked the `[-]` -> `collapse` change. But, in the interest of incremental progress, I think it's fine to discuss that later.

GuillaumeGomez commented:
I liked it as well, maybe for another time...

jsha commented:
@GuillaumeGomez I've added the tests.

GuillaumeGomez commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -312,15 +312,15 @@ impl AllTypes {
-                     </span>
-                 </span>
-             </h1>",
+             </h1>\
```
Why did you change this? With this change it looks like this:

![Screenshot from 2022-01-07 21-49-05](https://user-images.githubusercontent.com/3050060/148605576-63343dd3-aaf3-4444-acdf-da03493bbf7b.png)

jsha replied:
Good catch! This was a mistake from an earlier iteration. Reverted.

GuillaumeGomez commented:
Thanks for the tests! Just one last question and it's good for me. :)

GuillaumeGomez commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -315,8 +315,8 @@ impl AllTypes {
                          <a id=\"toggle-all-docs\" href=\"javascript:void(0)\" \
-                            title=\"collapse all docs\">\
-                             [<span class=\"inner\">&#x2212;</span>]\
+                             title=\"collapse all docs\">\
```
Why these changes?

GuillaumeGomez commented:
When I hover the source links, they don't get underlined. Could fix it and add a test for it too please?

jsha commented:
> When I hover the source links, they don't get underlined. Could fix it and add a test for it too please?

The source link at the top of the page is part of the main-heading, where links get underline on hover.

The source link to the right of the code headings is part of those headings, which don't get underline on hover.

I think this PR is consistent with how we currently do things. If we want to add underline-on-hover everywhere, I'm in favor, but we shouldn't do it piecemeal.

I pushed to remove the spurious diffs. Thanks for catching!

GuillaumeGomez commented:
It's fine as is then. I think it's worth discussing for a follow-up which links should be underlined or not but it's for later. r=me once CI pass and thanks!

jsha commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 962c0a4ee59e3c1a2413e785694c9433ffd0a9e1 has been approved by `jsha`

<!-- @bors r=jsha 962c0a4ee59e3c1a2413e785694c9433ffd0a9e1 -->
<!-- homu: {"type":"Approved","sha":"962c0a4ee59e3c1a2413e785694c9433ffd0a9e1","approver":"jsha"} -->

jsha commented:
@bors r=GuillaumeGomez

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #92534

bors commented:
:pushpin: Commit 962c0a4ee59e3c1a2413e785694c9433ffd0a9e1 has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez 962c0a4ee59e3c1a2413e785694c9433ffd0a9e1 -->
<!-- homu: {"type":"Approved","sha":"962c0a4ee59e3c1a2413e785694c9433ffd0a9e1","approver":"GuillaumeGomez"} -->
